File: voidfindertk/svf_popcorn/popcorn_void_finder/Example/read_voids.r
#--------------------------------------------------------------------------
# Spherical reader

read_sph <- function(filename) {
   
  sphvoids <- read.table(filename, header=FALSE)
  colnames(sphvoids) <- c("id", "r", "x", "y", "z", "delta")

  nsph <- length(sphvoids$id)
  print(c("Number of spheres: ", nsph))

  return(sphvoids)

}

#--------------------------------------------------------------------------
# Popcorn reader

read_pop <- function(filename){

    con  <- file(filename, open="r")
    npop <- as.numeric(scan(con, what="", nlines=1, quiet=TRUE)) # an integer
    print(c("Number of popcorns: ", npop))

    for(i in 1:npop){

        head_pop <- as.numeric(scan(con, what="", nlines=1, quiet=TRUE)) # line that contains 5 numbers
        id    = head_pop[1] # an integer
        nmem  = head_pop[2] # an integer
        vol   = head_pop[3] # a float
        npart = head_pop[4] # an integer
        flag  = head_pop[5] # an integer

        x=NULL; y=NULL; z=NULL; r=NULL; fvol=NULL; level=NULL
        if(nmem>0){
            for(j in 1:nmem){
                popmem <- as.numeric(scan(con, what="", nlines=1, quiet=TRUE)) # line that contains 6 numbers
                x = c(x,popmem[1]) # adding a float
                y = c(y,popmem[2]) # adding a float
                z = c(z,popmem[3]) # adding a float
                r = c(r,popmem[4]) # adding a float
                fvol = c(fvol,popmem[5]) # adding a float
                level = c(level,popmem[6]) # adding an integer	
            }
        }

        if(npart>0){
            for(j in 1:npart){
                ttt <- as.numeric(scan(con, what="", nlines=1, quiet=TRUE))
		# integers that are not stored 
            }
        }
 
        pop = data.frame(x=x, y=y, z=z, r=r)
        df = list(id=id, nmem=nmem, vol=vol, npart=npart, pop=pop)

        if(i==1){
          popcorn=list(df)
        } else {
          popcorn[[i]]=df
        }

    }

    close(con)
    return(popcorn)

}

#--------------------------------------------------------------------------
# Effective popcorn spheres

pop2sph <- function(popcorn){

    npop <- length(popcorn)
    print(c("Number of pop-spheres: ", npop))
    
    id = NULL; x = NULL; y = NULL; z = NULL; r = NULL; vol = NULL
    for(i in 1:npop){
        id  <- c(id,  popcorn[[i]]$id)
        x   <- c(x,   popcorn[[i]]$pop$x[1])
        y   <- c(y,   popcorn[[i]]$pop$y[1])
        z   <- c(z,   popcorn[[i]]$pop$z[1])
        r   <- c(r,   popcorn[[i]]$pop$r[1])
        vol <- c(vol, popcorn[[i]]$vol)
        
        reff <- (vol*3/(4*pi))**(1/3)
    }
    
    effvoids <- data.frame(id, reff, x, y, z)
    colnames(effvoids) <- c("id", "r", "x", "y", "z")

    return(effvoids)

}

File: voidfindertk/svf_popcorn/popcorn_void_finder/Example/read_voids.py
import numpy as np

#--------------------------------------------------------------------------
# Spherical reader

def read_sph(filename):
    """
    Reads a data table of spherical voids from an ASCII file (without header) and structures it as a dictionary.

    Parameters:
    - filename (str): the path to the data file.

    Returns:
    - sphvoids (dict): a dictionary with keys corresponding to spherical voids attributes.

    Spherical voids attribute:
    * id: void ID (integer)
    * r: radii of the member spheres (float)
    * x: x-coordinate of the centres of the member spheres (float)
    * y: y-coordinate (float)
    * z: z-coordinate (float)
    * delta: integrated density contrast at void radius (float)
    """

    file = np.loadtxt(filename)

    nsph = len(file)
    print(f"Number of spheres: {nsph}")

    columns = ["id", "r", "x", "y", "z", "delta"]
    sphvoids = {col: file[:,i] for i, col in enumerate(columns)}
    
    return sphvoids

#--------------------------------------------------------------------------
# Popcorn reader

def read_pop(filename):
    """
    Reads a data table of popcorn voids from an ASCII file (without header) and structures it as a dictionary.

    Parameters:
    - filename (str): The path to the data file.

    Returns:
    - popcorn (dict): a dictionary with keys corresponding to popcorn voids attributes.

    Popcorn voids attribute:
    * id: popcorn ID (integer)
    * nmem: number of member spheres (integer)
    * vol: popcorn volume (float)
    * reff: effective radius (float)
    * npart: number of particles inside (integer)
    * flag: for internal control, you can ignore it... (integer)
    * pop: popcorn object with the attributes of the member spheres (dictionary, length: nmem)
        * x: x-coordinate of the centres of the member spheres (float)
        * y: y-coordinate (float)
        * z: z-coordinate (float)
        * r: radii of the member spheres (float)
        * fvol: volume contribution to the total volume (float)
        * level: hierarchy level: 0 for main sphere, 1 for secondary spheres, ... (integer)
    """

    with open(filename, 'r') as file:
        npop = int(file.readline().strip()) # number of popcorn voids (integer)
        print(f"Number of popcorns: {npop}")
        
        popcorn = {'id':[], 'nmem':[], 'vol':[], 'reff':[], 'npart':[], 'flag':[], 'pop':[]}

        for p in range(npop):
            head_pop = list(file.readline().strip().split()) # this line is the header
            id    = int(head_pop[0])
            nmem  = int(head_pop[1])
            vol   = float(head_pop[2])
            npart = int(head_pop[3])
            flag  = int(head_pop[4])

            reff = (vol*3/(4*np.pi))**(1./3.)
	    # NOTE: this is a derived result, not contained in the original catalogue

            # Building the `pop` object:
            x=[]; y=[]; z=[]; r=[]; fvol=[]; level=[]
            if nmem > 0:
                for _ in range(nmem):
                    popmem = list(file.readline().strip().split()) # attributes of the member spheres
                    x.append(float(popmem[0]))
                    y.append(float(popmem[1]))
                    z.append(float(popmem[2]))
                    r.append(float(popmem[3]))
                    fvol.append(float(popmem[4]))
                    level.append(int(popmem[5]))
            pop = {'x':x, 'y':y, 'z':z, 'r':r, 'fvol':fvol, 'level':level} # this is an individual popcorn

            if npart > 0:
                for _ in range(npart):
                    file.readline() # reading inner particles ID (ignoring)

            popcorn['id'].append(id)
            popcorn['nmem'].append(nmem)
            popcorn['vol'].append(vol)
            popcorn['reff'].append(reff)
            popcorn['npart'].append(npart)
            popcorn['flag'].append(flag)
            popcorn['pop'].append(pop)

    return popcorn
